torch/csrc/api/src/optim/lbfgs.cpp

```diff
@@ -30,139 +68,191 @@ Tensor LBFGS::gather_flat_grad() {
       views.push_back(parameter.grad().view(-1));
     }
   }
-  return torch::cat(views);
+  return torch::cat(views, 0);
 }
 
-void LBFGS::add_grad(const torch::Tensor& step_size, const Tensor& update) {
+int64_t LBFGS::_numel() {
+  if(_numel_cache == c10::nullopt) {
+    auto res = 0;
+    for (auto& parameter : _params) {
+      int64_t numel = parameter.numel();
+      res+=numel;
+    }
+    *_numel_cache = res;
+  }
+  return *_numel_cache;
+}
+
+void LBFGS::_add_grad(const double step_size, const Tensor& update) {
   NoGradGuard guard;
   int64_t offset = 0;
-  for (auto& parameter : parameters_) {
+  for (auto& parameter : _params) {
     int64_t numel = parameter.numel();
     parameter.add_(
-        update.slice(0, offset, offset + numel, 1).view_as(parameter),
-        step_size.item<float>());
+      update.slice(0, offset, offset + numel, 1).view_as(parameter), step_size);
```

Comment:
It would be awesome to add the comment above as well: `// view as to avoid deprecated pointwise semantics`